CardaxDEX purs-bridge - testnet launch work PR

gnumonik opened the pull request:
This branch contains fixes to the codegen/printer which should ensure compatibility with Plutus Ledger types. The only errors I'm getting are missing Encode/DecodeJson instances for `BigInt` and `PlutusData` (which are expected, that's why we need `EncodeAeson`/`DecodeAeson`) 

I hate pretty printers.

Commit:
deferred newtype rec field erasure

Commit:
blah revert this

Commit:
revert this

Commit:
stupid commit to satisfy make plutus-ledger-api-typelib

Commit:
forgot to commit everything after running the formatter

bladyjoker commented:
TODO:
- [x] Resolve Argonaut vs PlutusTx integer conflict (Argonaut doesn't handle Integers, and PlutusTx doesn't handle Ints)
  - **STATUS**: Vladimir is trying to make CTL use Aeson solely so we can print into that?
  - [x] https://github.com/mlabs-haskell/purescript-bridge-aeson-helpers/pull/1/
- [x] Specify Arbitrary for plutus-ledger-api types
- [x] Embed plutus-ledger-api types into testing types
- [x] Enable testing in CI

NOTE:
- Passing PlutusData compatibility tests since https://github.com/mlabs-haskell/purescript-bridge/pull/21/commits/ca2e2a629192aa2482ca6913f6843b8d0f7041c5
- Json compatibility tests are running but failing since [b77d809](https://github.com/mlabs-haskell/purescript-bridge/pull/21/commits/b77d809421b07392af4423cfe3a8f82280a5f0bd)

TODO 2 (cc @gnumonik ):
- [x] Fix Aeson instances in CTL using PR https://github.com/Plutonomicon/cardano-transaction-lib/pull/419
    - Setup your packages.dhall to use local CTL Spago project in (just add // overrides)
- [ ] Add Rationals to unblock @MangoIV in https://github.com/CardaxDEX/blockchain-contracts/pull/350#issuecomment-1129942101
- [ ] Fix type def duplication bug in https://github.com/CardaxDEX/blockchain-contracts/pull/355 to unblock @Guilherme775

bladyjoker committed:
Merge branch 'master' into newtype-field-fixes

bladyjoker committed:
Point to mlabs-haskell/purescript-bridge-aeson-helpers

bladyjoker committed:
Synced to Plutonomicon/cardano-transaction-lib/pull/417 (complains about encodeAeson')

bladyjoker committed:
Synced with Plutonomicon/cardano-transaction-lib/pull/419 and it works

MangoIV committed:
fix nix:
- add the required packages to nix and cabal
- add .direnv/ to .gitignore

Commit:
Implemented arbitrary instances for all of the Ledger types except Validator/MintingPolicy, implement a sum type wrapper for those Ledger types to facilitate testing

Commit:
added plutus-extra as a dependency in cabal. also disabled -Werror temporarily for my sanity.

bladyjoker committed:
Small fix to haskell.nix

bladyjoker committed:
RT test works for Value and AssocMap and Interval (fails on Datum)

Commit:
special eq instance for tests, pretty print test output. i've run this a bunch of times and we're only getting the purescript bytestring parse error presumably caused by invalid maps, even w/ the tests turned on for all of the TestPlutusData types

Commit:
formatting/linting

Commit:
generic sorcery to replace broken Arbitrary PlutusData Maps with not-broken Arbitrary AssocMap Maps

Commit:
documented the confusing type classes/functions. hopefully coherently?

Commit:
tests all pass if the size is ~15 or less. small increases lead to exponential growth in the size of the types generated.

bladyjoker committed:
Cosmetic fixes

bladyjoker committed:
Cosmetics

bladyjoker committed:
Fixed Makefile

bladyjoker committed:
s/INCOHERENT/OVERLAPPABLE

bladyjoker committed:
Unflake the plutus-extra input due to a build err

error: derivation '/nix/store/p61f2ji9kjkpmnarwj41nvyc7pdncawh-hoogle-with-packages.drv' requires non-existent output 'doc' from input derivation '/nix/store/02bvcjhr2hjyw1j7d1xigr1rjsbpz73r-flat-lib-flat-0.4.5-haddock.drv'

bladyjoker committed:
All ledger types have Haskell JSON instances and Purs Aeson instances (compiles)

bladyjoker@morphism:~/Desktop/purescript-bridge$ nix develop .#roundTripTest -c cabal run test:tests -- --match "/Round trip tests (Purescript <-> Haskell)/With plutus-ledger-api bridge/should produce Aeson-compatible representations/" --seed 197079059

hs > waitUntil> [info] Build succeeded.
hs > waitUntil> I was born ready
hs> Req RTJson "{\"tag\":\"PlutusData\",\"contents\":{\"tag\":\"PdValue\",\"contents\":{\"getValue\":[[{\"unCurrencySymbol\":\"510c195b12023a6f0c671e0c74226c250b22134d125932630a133a2d\"},[[{\"unTokenName\":\"\"},12],[{\"unTokenName\":\"\\u0001O\\u0006=^M\\n\\u001e\\u0012x\\t3/@\"},9],[{\"unTokenName\":\"\\u001c\\u0006\\\"\\u000c\\u0018h\\u0006\"},-6]]],[{\"unCurrencySymbol\":\"54236a0e7d7279221d2e7331650a1202512e721e7f1b7e110d627251\"},[[{\"unTokenName\":\"3z\\u0008geJ\\u0008>(.\"},8],[{\"unTokenName\":\"m,C\DELAa\\u0000~5\\\"\\u0002\DELtc\"},-7],[{\"unTokenName\":\"tDL\\u0004Ej(H\\u001d\\u0018b\\u0001\"},-14]]],[{\"unCurrencySymbol\":\"625d0b4b473c7364091c5f173916617a6a285b337a1e670d3531644b\"},[[{\"unTokenName\":\"\\u0018/BN^4\\u001a\\\\t\"},-15],[{\"unTokenName\":\"(FGV\\u00164;\\u0013n\"},-6],[{\"unTokenName\":\"Y\"},7]]]]}}}"
    should produce Aeson-compatible representations [✘]

Failures:

  test/RoundTrip/Spec.hs:186:7:
  1) Round trip tests (Purescript <-> Haskell), With plutus-ledger-api bridge, should produce Aeson-compatible representations
       Falsifiable (after 1 test):
         PlutusData (PdValue (Value (Map [(510c195b12023a6f0c671e0c74226c250b22134d125932630a133a2d,Map [("",12),("O=^M
         x      3/@",9),(""
                           h",-6)]),(54236a0e7d7279221d2e7331650a1202512e721e7f1b7e110d627251,Map [("3ge>(.",8),("m,CAa~5"tc",-7),("tDLEj(Hb",-14)]),(625d0b4b473c7364091c5f173916617a6a285b337a1e670d3531644b,Map [("/BN^4\t",-15),("(FGV4;n",-6),("Y",7)])])))
       hs> Purescript shouldn't report an error
       expected: ""
        but got: "ps> Wanted Json got err: An error occurred while decoding a JSON value:"

  To rerun use: --match "/Round trip tests (Purescript <-> Haskell)/With plutus-ledger-api bridge/should produce Aeson-compatible representations/"

Randomized with seed 197079059

Finished in 79.6221 seconds

bladyjoker commented:
@gnumonik you can start with the following test

```
bladyjoker@morphism:~/Desktop/purescript-bridge$ nix develop .#roundTripTest -c cabal run test:tests -- --match "/Round trip tests (Purescript <-> Haskell)/With plutus-ledger-api bridge/should produce Aeson-compatible representations/" --seed 197079059
```

As it fails on (I think) TokenName JSON reps

```
hs > waitUntil> [info] Build succeeded.
hs > waitUntil> I was born ready
hs> Req RTJson "{\"tag\":\"PlutusData\",\"contents\":{\"tag\":\"PdValue\",\"contents\":{\"getValue\":[[{\"unCurrencySymbol\":\"510c195b12023a6f0c671e0c74226c250b22134d125932630a133a2d\"},[[{\"unTokenName\":\"\"},12],[{\"unTokenName\":\"\\u0001O\\u0006=^M\\n\\u001e\\u0012x\\t3/@\"},9],[{\"unTokenName\":\"\\u001c\\u0006\\\"\\u000c\\u0018h\\u0006\"},-6]]],[{\"unCurrencySymbol\":\"54236a0e7d7279221d2e7331650a1202512e721e7f1b7e110d627251\"},[[{\"unTokenName\":\"3z\\u0008geJ\\u0008>(.\"},8],[{\"unTokenName\":\"m,C\DELAa\\u0000~5\\\"\\u0002\DELtc\"},-7],[{\"unTokenName\":\"tDL\\u0004Ej(H\\u001d\\u0018b\\u0001\"},-14]]],[{\"unCurrencySymbol\":\"625d0b4b473c7364091c5f173916617a6a285b337a1e670d3531644b\"},[[{\"unTokenName\":\"\\u0018/BN^4\\u001a\\\\t\"},-15],[{\"unTokenName\":\"(FGV\\u00164;\\u0013n\"},-6],[{\"unTokenName\":\"Y\"},7]]]]}}}"
    should produce Aeson-compatible representations [✘]    

Failures:

  test/RoundTrip/Spec.hs:186:7: 
  1) Round trip tests (Purescript <-> Haskell), With plutus-ledger-api bridge, should produce Aeson-compatible representations
       Falsifiable (after 1 test):
         PlutusData (PdValue (Value (Map [(510c195b12023a6f0c671e0c74226c250b22134d125932630a133a2d,Map [("",12),("O=^M
         x      3/@",9),(""
                           h",-6)]),(54236a0e7d7279221d2e7331650a1202512e721e7f1b7e110d627251,Map [("3ge>(.",8),("m,CAa~5"tc",-7),("tDLEj(Hb",-14)]),(625d0b4b473c7364091c5f173916617a6a285b337a1e670d3531644b,Map [("/BN^4\t",-15),("(FGV4;n",-6),("Y",7)])])))
       hs> Purescript shouldn't report an error
       expected: ""
        but got: "ps> Wanted Json got err: An error occurred while decoding a JSON value:"

  To rerun use: --match "/Round trip tests (Purescript <-> Haskell)/With plutus-ledger-api bridge/should produce Aeson-compatible representations/"
```

CTL expects hex base16 encoded bytestring, but Haskell sends JSON escaped string or whatever (https://github.com/input-output-hk/plutus/blob/f805ab164a2aaeccd43224158647862164f8b0af/plutus-ledger-api/src/Plutus/V1/Ledger/Value.hs#L100)

bladyjoker committed:
Productivty boost and sanity preservation 100000%

bladyjoker committed:
Bridged Rational

bladyjoker committed:
Upgraded Nix utils for generating bridged typelibs

bladyjoker committed:
Added spagoProjectDir so we can include as spagoLocals in other projects

bladyjoker committed:
Small fix

bladyjoker committed:
Debugging stuff

bladyjoker committed:
Remove debug

bladyjoker committed:
Fix a small bug s/src/generated

bladyjoker committed:
A bunch of UX improvements to `purescript-flake.nix`

bladyjoker committed:
Nixified `purescript-flake.nix` more by turning localsDhall into a drv

bladyjoker committed:
Small bugfix

bladyjoker committed:
Move away from using env vars in favor of `locals.dhall` link

The issue encountered:
[nix-develop cardax-v1 ~/Desktop/blockchain-contracts/cardax-dex-browser-lib]$ spago build
spago:
 ./spago.dhall
   ./packages.dhall
     env:LOCALS_DHALL
       /nix/store/wg67f2k6g3rbkg7976aklkj1c7sjzy1c-locals-dhall
         /nix/store/sdmhb56imsixianrmibv82v12z082fam-spago-project-dir/spago.dhall
           /nix/store/sdmhb56imsixianrmibv82v12z082fam-spago-project-dir/packages.dhall

Cyclic import: env:LOCALS_DHALL

5                 env:LOCALS_DHALL

/nix/store/sdmhb56imsixianrmibv82v12z082fam-spago-project-dir/packages.dhall:5:17

bladyjoker committed:
Another round of fixes (spago wants src/ and can't use generated/ fsr)

gnumonik commented:
Hey @bladyjoker I'm not sure if the issues are up to date and I need a bit of explanation for all of the changes you made since I finished last week. Ping me on slack for a quick meeting when you wake up.

bladyjoker committed:
Bump CTL

bladyjoker committed:
Small cosmetic fixes to purescript-flake.nix

MangoIV committed:
add check attr

bladyjoker committed:
Used `RespError` and added `localOverrides` in `packages.dhall`

hs> Req RTJson "{\"tag\":\"Either\",\"contents\":{\"Left\":true}}"
ps> RespSuccess RTJson "{\"contents\":{\"Left\":true},\"tag\":\"Either\"}"
hs> Req RTJson "{\"tag\":\"PlutusData\",\"contents\":{\"tag\":\"PdTypesWithMap\",\"contents\":{\"tag\":\"PdDatum\",\"contents\":\"21\"}}}"
    should produce Aeson-compatible representations [0/100]ps> RespError "ps> Wanted Json got err: (Named \"TestData\" (AtKey \"contents\" (Named \"TestPlutusData\" (AtKey \"contents\" (Named \"TypesWithMap\" (AtKey \"contents\" (TypeMismatch \"Object\"))))))) on input: {\"tag\":\"PlutusData\",\"contents\":{\"tag\":\"PdTypesWithMap\",\"contents\":{\"tag\":\"PdDatum\",\"contents\":\"21\"}}}"
    should produce Aeson-compatible representations [✘]

Failures:

  test/RoundTrip/Spec.hs:140:28:
  1) Round trip tests (Purescript <-> Haskell), With plutus-ledger-api bridge, should produce Aeson-compatible representations
       Falsifiable (after 2 tests):
         PlutusData (PdTypesWithMap (PdDatum (Datum {getDatum = I (-2)})))
       hs> Wanted ResSuccess got ResError ps> Wanted Json got err: (Named "TestData" (AtKey "contents" (Named "TestPlutusData" (AtKey "contents" (Named "TypesWithMap" (AtKey "contents" (TypeMismatch "Object"))))))) on input: {"tag":"PlutusData","contents":{"tag":"PdTypesWithMap","contents":{"tag":"PdDatum","contents":"21"}}}

  To rerun use: --match "/Round trip tests (Purescript <-> Haskell)/With plutus-ledger-api bridge/should produce Aeson-compatible representations/"

Randomized with seed 696976685

Finished in 3.1022 seconds
1 example, 1 failure

bladyjoker committed:
Synced CTL and small fixes/upgrades to `purs-lib.nix`

bladyjoker committed:
`localOverrides` must be commented out...

[nix-develop cardax-v1 ~/Desktop/blockchain-contracts/cardax-dex-browser-lib]$ make
rm -fR node_modules .spago2nix/ .spago output .psci_modules generated
spago2nix generate
getting packages..
Failed to list packages:
spago:
 ./spago.dhall
   ./packages.dhall
     ./locals.dhall
       /nix/store/wp3gh5i15qy7wh4f1y4fq82qnvldvk7d-purescript-lib-build-purs-project-spagoProjectDir/spago.dhall
         /nix/store/wp3gh5i15qy7wh4f1y4fq82qnvldvk7d-purescript-lib-build-purs-project-spagoProjectDir/packages.dhall
           /cardano-transaction-lib/spago.dhall

Error: Missing file /cardano-transaction-lib/spago.dhall

34            ../../../cardano-transaction-lib/spago.dhall

/nix/store/wp3gh5i15qy7wh4f1y4fq82qnvldvk7d-purescript-lib-build-purs-project-spagoProjectDir/packages.dhall:34:12

(SpagoRunError "Normally 1")

bladyjoker committed:
Another small fix to `purs-lib.nix` to include only `src` from 'locals'

bladyjoker committed:
Forgot to sync with latest CTL that has Rational Rings

Commit:
fixed datum json instance codgen

Commit:
Merge branch 'newtype-field-fixes' of github.com:mlabs-haskell/purescript-bridge into newtype-field-fixes

MangoIV committed:
ignore locals.dhall

MangoIV committed:
Merge branch 'mangoiv/try-hydra' into newtype-field-fixes

MangoIV committed:
update ctl

MangoIV committed:
update ctl (this time really)

bladyjoker committed:
Small debugging fix

bladyjoker committed:
Sync with CTL that has address conversion fixes

bladyjoker committed:
Sync with fixed CTL

Commit:
updated ctl version

MangoIV committed:
updated ctl

bladyjoker committed:
CTL sync

bladyjoker committed:
Merge branch 'newtype-field-fixes' of github.com:mlabs-haskell/purescript-bridge into newtype-field-fixes

bladyjoker committed:
Update documentation on how to reference local CTL repo for quicker cross repo work

bladyjoker committed:
Refactored CLI handling

bladyjoker committed:
Nixified the round trip test and did a big reshuffle

bladyjoker committed:
Synced with latest CTL

bladyjoker committed:
Fix node deps

klntsky committed:
Epic fix: generate correct EncodeAeson instances

bladyjoker committed:
Merge pull request #24 from klntsky/klntsky/epic-fix

Fix generated EncodeAeson instances

bladyjoker committed:
Reshuffle and fix haskell-language-server

bladyjoker committed:
Fix codegen unit tests

bladyjoker committed:
Fix build/lint

bladyjoker committed:
`Completed (22 files worked, 0 files failed)` HLS flawless

MangoIV committed:
updated ctl, tests pass

Commit:
Started writing documentation

Commit:
Added a function that allows users to specify their own bridge parts when writing ledger types

gnumonik committed:
Merge branch 'newtype-field-fixes' of github.com:mlabs-haskell/purescript-bridge into newtype-field-fixes

gnumonik committed:
Finished initial draft of readme

MangoIV committed:
test pr (#27)

build with hydra, deprecate gh workflows

gnumonik committed:
fixed makefile, re-generated ledger-api-typelib

gnumonik committed:
Merge branch 'newtype-field-fixes' of github.com:mlabs-haskell/purescript-bridge into newtype-field-fixes

gnumonik committed:
updated generated folder location in flake

gnumonik committed:
formatting
